Add core profiler "Welcome to Woo!" page

chihsuan opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/team-ghidorah/issues/195.

figma: Y5pUYSJPsGEud1vknUZhi8-fi-739_55969

![00  Welcome to Woo](https://user-images.githubusercontent.com/4344253/234208217-3e530610-6a05-413a-85ff-eeeef12b74a6.png)

<img src="https://user-images.githubusercontent.com/4344253/234212410-4e49966b-323e-49d8-af6b-7c55aeb29c10.png" width="300px" />


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Scenario: User enters the OBW**

1. Install and activate WooCommerce on a brand new site
2. Install the `WooCommerce Beta Tester` plugin
3. Go to /wp-admin/tools.php?page=woocommerce-admin-test-helper
4. Enable the 'core-profiler' feature flag
5. Visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard
6. Confirm the UI matches the design. (Y5pUYSJPsGEud1vknUZhi8-fi-739_55969)
7. Confirm that the allow-tracking checkbox is checked.
8. Open your browser developer console and run `localStorage.setItem( 'debug', 'wc-admin:*' );` to set debug for tracks
9. Update woocommerce_allow_tracking option to true
10. Refresh the page
11. Open your browser developer console and confirm `storeprofiler_step_view` track is recorded.
12. Update `woocommerce_allow_tracking` option to false
13. Visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard
14. Observe that the allow-tracking checkbox is unchecked.

**Scenario: User clicks on the Skip guided setup button**

1. Visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard
2. Clicks on the `Skip guided setup`
3. Confirmed `storeprofiler_store_details_skip` track is recorded
4. Observe that you're redirected to  step User Profile

**User clicks on the Set up my store**

1. Visit http://localhost:8888/wp-admin/admin.php?page=wc-admin&path=%2Fsetup-wizard
2. Clicks on the `Set up my store` button
3. Confirmed `storeprofiler_step_complete` track is recorded
4. Observe that you're redirected to `Business Location` step

![localhost_8888_wp-admin_admin php_page=wc-admin path=%2Fsetup-wizard doing_wp_cron=1682414823 1799008846282958984375](https://user-images.githubusercontent.com/4344253/234235629-ee75908b-287b-478d-b74f-568f6b6c68cb.png)

<img src="https://user-images.githubusercontent.com/4344253/234235697-ac6c4543-3d9a-4615-8cea-f6f24a8a8db4.png" width="300px" />

<!-- End testing instructions -->

chihsuan committed:
Add core profiler - welcome to woo page

chihsuan committed:
Add changelog

chihsuan committed:
Update checkbox styles

chihsuan committed:
Update copies

chihsuan committed:
Add test

chihsuan committed:
Add enable prop to window.wcTracks type

chihsuan committed:
Add tracks

chihsuan committed:
Add explat changelog

chihsuan committed:
Fix import

chihsuan committed:
Update default tracking value

chihsuan committed:
Update copies

chihsuan committed:
Fix test

chihsuan commented on plugins/woocommerce-admin/client/core-profiler/index.tsx:
```diff
@@ -268,26 +308,62 @@ const coreProfilerStateMachineDefinition = createMachine( {
 	const [ state, send ] = useMachine(
 		coreProfilerStateMachineDefinition.withConfig( {
-			services: {
-				initializeProfiler: () => ( sendBack ) => {
```
@rjchow: I removed the `initializeProfiler` and [sync the data with `useEffect`](https://xstate.js.org/docs/recipes/react.html#other-hooks) instead because I need to fetch the `woocommerce_allow_tracking` option using `useSelect`. I'm not sure if there is a better way, any feedback is highly welcome!

rjchow replied:
updated!

moon0326 reviewed (dismissed):
Looks great!! Thank you for working on it so quickly 👍 🚀

rjchow committed:
using invoked promise instead of useState

- take advantage of xstate's built ins for side effects instead of  useEffect/hooks
- discovered that error result wasn't really handled in original useEffect
- use text labels instead of inline functions so that we can decouple the implementation from the machine model
- todo: can move the invoked function out elsewhere and also tests if needed (not necessary here because it's a simple call)

rjchow committed:
use actions.choose instead of branching inside action 

- https://stately.ai/docs/xstate/transitions-and-choices/guarded-actions#the-choose-action
- makes it so that the machine model is serializeable
- todo: in the distant future i wonder if it might be tidier to have tracks be its own actor that just receives the same events that this machine does, that way it's just standalone instead of mixing up tracks with the implementation

rjchow committed:
use dispatch instead of useDispatch

- decouples the implementation of the handler from the react component
- makes the handler testable on its own if needed
- makes the state machine testable without relying on external dependencies

rjchow committed:
decoupled remaining function calls

rjchow commented on plugins/woocommerce-admin/client/core-profiler/index.tsx:
```diff
@@ -100,6 +110,65 @@ const Extensions = ( {
+	event: IntroOptInEvent
+) => {
+	if (
+		event.payload.optInDataSharing &&
+		typeof window.wcTracks.enable === 'function'
+	) {
+		window.wcTracks.enable( () => {
+			initializeExPlat();
+		} );
+	} else if ( ! event.payload.optInDataSharing ) {
+		window.wcTracks.isEnabled = false;
+	}
```
I'm unfamiliar with this bit, when do we need to initialize explat?

chihsuan replied:
This function is copied from the existing OBW code. `initializeExPlat` was added in https://github.com/woocommerce/woocommerce-admin/pull/6669/. I think we initialize here in case we need to do any explat experiments.

https://github.com/woocommerce/woocommerce/blob/64dbeba26072f66e7f8f52e65aabc02161c21ba9/plugins/woocommerce-admin/client/profile-wizard/steps/usage-modal.js#L59-L83

Usually, we initialize explat when explat the package is loaded.

https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/explat/src/index.ts#L25-L31

rjchow commented on plugins/woocommerce-admin/client/core-profiler/index.tsx:
```diff
@@ -100,6 +110,65 @@ const Extensions = ( {
+		window.wcTracks.isEnabled = false;
+	}
+
+	const trackingValue = event.payload.optInDataSharing ? 'yes' : 'no';
```
If it is easier we can just use 'yes' 'no' in the context, it slipped my mind that it's a stringly value and not boolean when I was scaffolding 😅

chihsuan replied:
I think it’s fine either way because if it's a stringly value, we still need to convert it to boolean for checkbox. 😅

rjchow commented on plugins/woocommerce-admin/client/core-profiler/index.tsx:
```diff
@@ -100,6 +110,65 @@ const Extensions = ( {
 	);
 };
 
+const getAllowTrackingOption = async () =>
```
Do you think we should move these out into more specific collections instead of being located in this file?

chihsuan replied:
Yeah, that would be great!

rjchow reviewed (commented):
Thanks Chi! This is beautiful!
I made some documented commits, have a look over please!

I think there's a mistake somewhere causing the useMachine changed between renders warning, I'll look into that separately tomorrow

chihsuan committed:
Fix lint error

chihsuan commented:
Thank you @rjchow! The refactor looks great, it’s cleaner. 😍

chihsuan committed:
Fix style lint

rjchow committed:
address xstate console warnings

- .withConfig() is a function call that returns a new object every invocation so we need to wrap it in useMemo to keep it stable

rjchow reviewed (dismissed):
I think this is good to go!

adrianduffell reviewed (changes requested):
Nice work @chihsuan , this tested great for me!

One suggestion: could we bolster the testing for the consent checkbox?

It looks like there's a test for `optInDataSharing is true`. It would also be good to test the `false` condition too -- it would be bad to have a future regression there!

chihsuan committed:
Add optInDataSharing is false test

chihsuan commented:
> Nice work @chihsuan , this tested great for me!
> 
> One suggestion: could we bolster the testing for the consent checkbox?
> 
> It looks like there's a test for `optInDataSharing is true`. It would also be good to test the `false` condition too -- it would be bad to have a future regression there!

Sure thing! Added in 900abd18d6fe2a36dcf778f409195ac24ef1a8d3

adrianduffell reviewed (dismissed):
Thanks for adding the test! LGTM! 🚀

chihsuan committed:
Fix lint

chihsuan commented:
@adrianduffell or @rjchow 👋  Could I get an approval again on this PR, please? I pushed a lint fix commit. That..removed your approvals. 🥲

rjchow reviewed (approved):
😃
